superset/queries/saved_queries/dao.py

```diff
@@ -44,3 +47,20 @@ def bulk_delete(models: Optional[list[SavedQuery]], commit: bool = True) -> None
         except SQLAlchemyError as ex:
             db.session.rollback()
             raise DAODeleteFailedError() from ex
+
+    @classmethod
+    def get_by_id(cls, _id: str) -> Optional[SavedQuery]:
```

Comment:
I don't believe this is the right approach, i.e., per this code block the UUID option is bypass the security check (by way of the filter on line #63). The correct approach would be to generate a permalink similar to charts and dashboards.

cc @michael-s-molina and @villebro as we've discussed this previously.

Reply:
Bypassing the security check if a UUID is passed is intentional, it allows users to access saved queries via the API only if the UUID link has been shared with them.  There is precedence for this approach in the codebase, see the dashboard DAO: https://github.com/apache/superset/blob/master/superset/dashboards/dao.py#L47

Reply:
I would use the permalink logic for this purpose and probably migrate the dashboard DAO too. Keeping multiple strategies for sharing resources only increases code complexity and opens the door for inconsistent security checks.

Reply:
@michael-s-molina or @john-bodley given the discussion in the all hands meetup, did we come to a decision on whether we can move towards UUIDs here and save the permalink feature for a moment in time state that you want to share?

Reply:
@eschutho I’m not sure that the UUID seems correct. It feels like an internal representation that we’re now exposing to circumvent security, i.e., it feels akin to saying users enter a bar and present their ID for verification, but if you wave a UUID everything is fine.

In my opinion the precedence which was set previously (the example @jfrag1 linked to—which is the sam one I was referring to in the town hall) should be thought of as the exception rather than the rule, i.e., that logic would also need to be updated in a future PR.

Reply:
If we think we need to change the way we identify resources because of security or portability concerns, we need to write a SIP about it because it will greatly impact the application (URL representation, APIs, storage, etc). We also need to check if UUIDs meet the security criteria or if we need a more secure approach that is also portable.

UUIDs are merely resource identifiers, similar to a sequential ID. They shouldn't have a different meaning in terms of security like @john-bodley mentioned above.

Reply:
I strongly disagree with the idea that there's no justification for UUID's being treated differently for security.  
> Granted it’s harder to forge a UUID as opposed to guess a valid ID number, but that still doesn’t seem secure.

Guesswork has nothing to do with it; sequential ID's are vulnerable to enumeration, meaning a user of the API could easily get access to all saved queries. This is the case with the old API with no security filter. Consider a script like the following:
```python
for i in range(1000):
   res = requests.get(f'savedqueryapi/get/{i}')
   # save response data somewhere
```
UUID's on the other hand, aren't "harder" to guess, they're _impossible_ to guess.  In the v1 API, the only way to obtain a saved query's UUID is by being logged in as the user who created the saved query, so the contents of the saved query are as secure as the user keeps the UUID/link to it.  I'd argue that this was the intent that saved queries were created with.  It's the way they've worked until [this change](https://github.com/apache/superset/pull/21682/files#diff-1fe9f23910bda67f66ff7b5e9127cdaba0d304153a5246dbcb2d76d70d2e5a4fR1365) was made. There's a "Copy link" button prominently in the UI, which seems to suggest that the idea is that you can then send that link to someone else to view.

While yes, the old API which enabled this functionality was not secure because it allowed enumeration, it was still a great feature to use.  I believe this PR brings the saved query feature back to what it was originally supposed to be in a much more secure way than before.

Reply:
I agree with @michael-s-molina's comment regarding making this a SIP. I understand that this feels somewhat like a blocker to what feels like a somewhat benign fix, but I think it's important that we flush out (and detangle) the key concepts that Michael and myself raised to set the right precedence, as opposed to a bandaid solution.

Historically UUIDs were obfuscated and solely used to make charts, dashboards, etc. globally unique across shared Superset instances, however their use within the API is more of an organic phenomenon and thus I think there's merit in pausing, taking a step back, and evaluating how we think about UUIDs, shareable links, etc. in a more holistic manner.

I agree that finding a matching UUID is near impossible, but using UUIDs for security (via cryptography) isn't valid. Note this is in no way a criticism of your work, but more with the state of the Superset code base. At Airbnb we support thousands of weekly active users and are constantly fixing issues related to problematic features and/or inconsistent/non-coherent/complex functionality (you can see the [vast majority](https://github.com/apache/superset/pulls/john-bodley) of my commits to this repo are actually fixes and/or mechanism to improve quality).

Reply:
Agree that a SIP on this functionality is a good idea. We were planning to write one up, I think @john-bodley is also interested in writing it. If we think this PR is too premature at this point, what do you all think about us reverting [this one line](https://github.com/apache/superset/pull/21682/files#diff-1fe9f23910bda67f66ff7b5e9127cdaba0d304153a5246dbcb2d76d70d2e5a4fR1365) back to the old api so that we can fix the sharing capabilities that are currently broken, and then we can talk about a longer-term solution in an upcoming SIP? cc @hughhhh who is also helping on this.

Reply:
@zephyring, @jfrag1, @eschutho, et al. I'm supportive of "reverting" the original PR to get us back to square one and then with said SIP we can proceed. I suspect the old API endpoint is a legacy [FAB REST API](https://flask-appbuilder.readthedocs.io/en/latest/quickhowto.html#rest-api) which was slated for deletion in 2.3.X (we're now using FAB 4.3.2+).

If the one line revert to the old API isn't via I think the alternative is something akin to what we do at Airbnb, where our PR description states:

> Though this logic,
>
>
> ```python
> class SavedQueryFilter(BaseFilter):  # pylint: disable=too-few-public-methods
>    def apply(self, query: BaseQuery, value: Any) -> BaseQuery:
>        """
>        Filter saved queries to only those created by current user.
>
>        :returns: flask-sqlalchemy query
>        """
>        
>        return query.filter(
>            SavedQuery.created_by == g.user  # pylint: disable=comparison-with-callable
>        )
>```
>which, restricts only the creator of the saved query to access it, has been around for [years](https://github.com/apache/superset/pull/10793) we suspect it only came into force when the /api/v1/saved_query/{pk} API was enabled in Q4 of 2022 which prevented people from sharing saved queries.
> 
> This PR simply relaxes the filter to allow anyone to access a saved query. 

Thus I _think_ an alternative to changing the API endpoint would be to simply replace [these lines](https://github.com/apache/superset/blob/884a8b52d6a8535ae54c7dab3bfcad3388658877/superset/queries/saved_queries/filters.py#L78-L80) with `return query`.

Reply:
opened https://github.com/apache/superset/pull/24434 to revert the original change which broke this functionality